ft-fine-tune/fine-tune.py

```diff
@@ -0,0 +1,206 @@
+DATASET_ENCODING = "ISO-8859-1"
+
+# Read data
+data = pd.read_csv(input_filename, encoding=DATASET_ENCODING, names=DATASET_COLUMNS)
+
+# decode_map = {0: "NEGATIVE", 2: "NEUTRAL", 4: "POSITIVE"}
+decode_map = {0: 0, 2: 1, 4: 2}
+def decode_sentiment(label):
+    return decode_map[int(label)]
+
+
+data.target = data.target.apply(lambda x: decode_sentiment(x))
+
+
+# TEXT CLENAING
+TEXT_CLEANING_RE = "@\S+|https?:\S+|http?:\S|[^A-Za-z0-9]+"
+
+def preprocess(text, stem=False):
+    # Remove link,user and special characters
+    text = re.sub(TEXT_CLEANING_RE, ' ', str(text).lower()).strip()
+    tokens = []
+    for token in text.split():
+        if stem:
+            tokens.append(stemmer.stem(token))
+        else:
+            tokens.append(token)
+    return " ".join(tokens)
+
+
+data.text = data.text.apply(lambda x: preprocess(x))
+data.head(15)
+
+# Define pretrained tokenizer and model
+model_name = "bert-base-uncased"
+current_path = os.getcwd()
+tokenizer = BertTokenizer.from_pretrained(model_name)
+model = BertForSequenceClassification.from_pretrained(model_name, num_labels=3).to(device)
+data.head()
+
+# from torchinfo import summary
+# summary(model)
+
+# Get all of the model's parameters as a list of tuples.
+params = list(model.named_parameters())
+
+print('The BERT model has {:} different named parameters.\n'.format(len(params)))
+
+print('==== Embedding Layer ====\n')
+
+for p in params[0:5]:
+    print("{:<55} {:>12}".format(p[0], str(tuple(p[1].size()))))
+
+print('\n==== First Transformer ====\n')
+
+for p in params[5:21]:
+    print("{:<55} {:>12}".format(p[0], str(tuple(p[1].size()))))
+
+print('\n==== Output Layer ====\n')
+
+for p in params[-4:]:
+    print("{:<55} {:>12}".format(p[0], str(tuple(p[1].size()))))
+##############Finish Printing Model Structure################
+
+# ----- 1. Preprocess data -----#
+# Preprocess data
+X = list(data[text_column].astype(str))
+y = list(data[label_column].astype(int))
+print(X[0:15])
+print(y[0:15])
+X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2)
+print(X_train[0:15])
+print(X_val[0:15])
+X_train_tokenized = tokenizer(X_train, padding=True, truncation=True, max_length=max_length, return_tensors='pt')
+X_val_tokenized = tokenizer(X_val, padding=True, truncation=True, max_length=max_length, return_tensors='pt')
+
+# Create torch dataset
+class Dataset(torch.utils.data.Dataset):
+    def __init__(self, encodings, labels=None):
+        self.encodings = encodings
+        self.labels = labels
+
+    def __getitem__(self, idx):
+        item = {key: val[idx] for key, val in self.encodings.items()}
+        item['labels'] = self.labels[idx]
+        print(item)
+        return item
+
+    def __len__(self):
+        return len(self.encodings["input_ids"])
+
+train_dataset = Dataset(X_train_tokenized, y_train)
+val_dataset = Dataset(X_val_tokenized, y_val)
+
+# ----- 2. Fine-tune pretrained model -----#
+# Define Trainer parameters
+def compute_metrics(p):
```

Comment:
move above the code flow. 

Also comment describing the arguments passed, operations and return values.